go-runtime/encoding/encoding.go

```diff
@@ -278,6 +268,63 @@ func decodeValue(d *json.Decoder, v reflect.Value) error {
 	}
 }
 
+func handleIfNextTokenIsNull(d *json.Decoder, ifNullFn func(*json.Decoder) error, elseFn func(*json.Decoder) error) error {
+	isNull, err := isNextTokenNull(d)
+	if err != nil {
+		return err
+	}
+	if isNull {
+		err = ifNullFn(d)
+		if err != nil {
+			return err
+		}
+		// Consume the null token
+		_, err := d.Token()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	return elseFn(d)
+}
+
+// isNextTokenNull implements a cheap/dirty version of `Peek()`, which json.Decoder does
+// not support.
+func isNextTokenNull(d *json.Decoder) (bool, error) {
```

Comment:
This feels pretty hacky. I'd be tempted to write a thin layer over `json.Decoder` that supports peeking, rather than doing this.

Reply:
It is very very gross. We spent a bit of time working on wrapping `json.Decoder`, but the issue is that we can't just wrap `Token()` (returning the last `Peek` value when defined instead of what's at the top of the buffer) and add `Peek` because `d.Decode` will pop tokens off the underlying buffer, and if we've popped a token into the wrapper object, the underlying Decoder's `Decode()` function won't use that. So unless there's an alternative I haven't thought of, we'd effectively have to reimplement the whole `json.Decoder`. Do you have any ideas?

Reply:
Ah right, that makes sense.

I think we want to ditch `json.Decoder` altogether, but that's not appropriate for this PR.